CombinatorialGames/Game/Computability/FGame.lean

```diff
@@ -0,0 +1,418 @@
+    ext z
+    simp_rw [Finset.mem_image, Finset.mem_univ, true_and]
+    constructor <;> rintro ⟨i, rfl⟩
+    · obtain ⟨j, hj⟩ := h.moveRight i
+      exact ⟨j, hj.mk_eq.symm⟩
+    · obtain ⟨j, hj⟩ := h.moveRight_symm i
+      exact ⟨j, hj.mk_eq⟩
+
+@[simp] theorem leftMoves_mk (x : SGame) :
+    leftMoves (mk x) = Finset.univ.image (mk ∘ x.moveLeft) := rfl
+@[simp] theorem rightMoves_mk (x : SGame) :
+    rightMoves (mk x) = Finset.univ.image (mk ∘ x.moveRight) := rfl
+
+@[ext]
+theorem ext {x y : FGame} (hl : x.leftMoves = y.leftMoves) (hr : x.rightMoves = y.rightMoves) :
+    x = y := by
+  cases x with | H x =>
+  cases y with | H y =>
+  dsimp at hl hr
+  refine (SGame.identical_iff.2 ⟨⟨?_, ?_⟩, ⟨?_, ?_⟩⟩).mk_eq <;> intro i
+  · have := Finset.mem_image_of_mem mk (show x.moveLeft i ∈ Finset.univ.image x.moveLeft by simp)
+    rw [Finset.image_image, hl] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb.symm⟩
+  · have := Finset.mem_image_of_mem mk (show y.moveLeft i ∈ Finset.univ.image y.moveLeft by simp)
+    rw [Finset.image_image, ← hl] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb⟩
+  · have := Finset.mem_image_of_mem mk (show x.moveRight i ∈ Finset.univ.image x.moveRight by simp)
+    rw [Finset.image_image, hr] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb.symm⟩
+  · have := Finset.mem_image_of_mem mk (show y.moveRight i ∈ Finset.univ.image y.moveRight by simp)
+    rw [Finset.image_image, ← hr] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb⟩
+
+/-- `IsOption x y` means that `x` is either a left or a right move for `y`. -/
+@[aesop simp]
+def IsOption (x y : FGame) : Prop :=
+  x ∈ y.leftMoves ∨ x ∈ y.rightMoves
+
+theorem IsOption.of_mem_leftMoves {x y : FGame} : x ∈ y.leftMoves → IsOption x y := .inl
+theorem IsOption.of_mem_rightMoves {x y : FGame} : x ∈ y.rightMoves → IsOption x y := .inr
+
+theorem isOption_wf : WellFounded IsOption := by
+  refine ⟨ind fun x ↦ ?_⟩
+  induction x with
+  | mk x _ _ _ hl hr =>
+    constructor
+    rintro ⟨y⟩ (h | h)
+    · rw [leftMoves_mk, Finset.mem_image] at h
+      exact h.choose_spec.2 ▸ hl h.choose
+    · rw [rightMoves_mk, Finset.mem_image] at h
+      exact h.choose_spec.2 ▸ hr h.choose
+
+instance : IsWellFounded _ IsOption := ⟨isOption_wf⟩
+
+theorem IsOption.irrefl (x : FGame) : ¬ IsOption x x := _root_.irrefl x
+
+theorem self_notMem_leftMoves (x : FGame) : x ∉ x.leftMoves :=
+  fun hx ↦ IsOption.irrefl x (.of_mem_leftMoves hx)
+
+theorem self_notMem_rightMoves (x : FGame) : x ∉ x.rightMoves :=
+  fun hx ↦ IsOption.irrefl x (.of_mem_rightMoves hx)
+
+/-- Construct a `FGame` from its left and right lists. This is an auxiliary definition
+used to define the more general `FGame.ofFinsets`. -/
+def ofLists (s t : List FGame.{u}) : FGame.{u} :=
```

Comment:
Can this provide any computational advantage over `ofFinsets`? If so, maybe we should treat it as more than just an auxiliary definition!